File: aiwos/lib/api/notifications.ts
import { apiClient } from "./client";

export type NotificationApiResponse = {
  id: string;
  organization_id: string;
  user_id: string | null;
  type: string;
  title: string;
  body: string | null;
  entity_id: string | null;
  entity_type: string | null;
  is_read: boolean;
  created_at: string;
};

export const notificationApi = {
  list: (organization_id: string, skip = 0, limit = 50) =>
    apiClient
      .get<NotificationApiResponse[]>("/notifications", {
        params: { organization_id, skip, limit },
      })
      .then((r) => r.data),

  markRead: (id: string) =>
    apiClient
      .patch<NotificationApiResponse>(`/notifications/${id}/read`)
      .then((r) => r.data),

  markAllRead: (organization_id: string) =>
    apiClient
      .post<{ marked_read: number }>("/notifications/read-all", null, {
        params: { organization_id },
      })
      .then((r) => r.data),
};

File: aiwos/aiwos-backend/app/api/v1/endpoints/notifications.py
import uuid
from typing import List

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.notification import NotificationResponse
from app.services import notification_service

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=List[NotificationResponse])
async def list_notifications(
    organization_id: uuid.UUID,
    skip: int = 0,
    limit: int = 50,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> List:
    return await notification_service.list_notifications(
        db, organization_id, current_user.id, skip=skip, limit=limit
    )


@router.patch("/{notification_id}/read", response_model=NotificationResponse)
async def mark_notification_read(
    notification_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    notification = await notification_service.mark_read(db, notification_id)
    if notification is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found."
        )
    return notification


@router.post("/read-all")
async def mark_all_notifications_read(
    organization_id: uuid.UUID = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    count = await notification_service.mark_all_read(db, organization_id, current_user.id)
    return {"marked_read": count}
